modules/gapi/src/executor/gtbbexecutor.cpp

```diff
@@ -0,0 +1,445 @@
+};
+
+using root_t = std::unique_ptr<tbb::task, destroy_tbb_task>;
+
+root_t inline create_root(tbb::task_group_context& ctx) {
+    root_t  root{new (tbb::task::allocate_root(ctx)) tbb::empty_task};
+    root->set_ref_count(1); // required by wait_for_all, as it waits until counter drops to 1
+    return root;
+}
+
+std::size_t inline tg_context_traits() {
+    // Specify tbb::task_group_context::concurrent_wait in the traits to ask TBB scheduler not to change
+    // ref_count of the task we wait on (root) when wait is complete.
+    return tbb::task_group_context::default_traits | tbb::task_group_context::concurrent_wait;
+}
+
+} // namespace tasking
+
+namespace async {
+struct async_tasks_t {
+    std::atomic<size_t>         count {0};
+    std::condition_variable     cv;
+    std::mutex                  mtx;
+};
+
+enum class wake_tbb_master {
+   no,
+   yes
+};
+
+void inline wake_master(async_tasks_t& async_tasks, wake_tbb_master wake_master) {
+    // TODO: seems that this can be relaxed
+    auto active_async_tasks = --async_tasks.count;
+
+    if ((active_async_tasks == 0) || (wake_master == wake_tbb_master::yes)) {
+        // Was the last async task or asked to wake TBB master up(e.g. there are new TBB tasks to execute)
+        GAPI_ITT_AUTO_TRACE_GUARD(ittTbbUnlockMasterThread);
+        // While decrement of async_tasks_t::count is atomic, it might occur after the waiting
+        // thread has read its value but _before_ it actually starts waiting on the condition variable.
+        // So, lock acquire is needed to guarantee that current condition check (if any) in the waiting thread
+        // (possibly ran in parallel to async_tasks_t::count decrement above) is completed _before_ signal is issued.
+        // Therefore when notify_one is called, waiting thread is either sleeping on the condition variable or
+        // running a new check which is guaranteed to pick the new value and return from wait().
+
+        // There is no need to _hold_ the lock while signaling, only to acquire it.
+        std::unique_lock<std::mutex> {async_tasks.mtx};   // Acquire and release the lock.
+        async_tasks.cv.notify_one();
+    }
+}
+
+struct master_thread_sleep_lock_t
+{
+    struct sleep_unlock {
+       void operator()(async_tasks_t* t) const {
+          ASSERT(t);
+          wake_master(*t, wake_tbb_master::no);
+       }
+    };
+
+    std::unique_ptr<async_tasks_t, sleep_unlock>  guard;
+
+    master_thread_sleep_lock_t() = default;
+    master_thread_sleep_lock_t(async_tasks_t*  async_tasks_ptr) : guard(async_tasks_ptr) {
+        // TODO: seems that this can be relaxed
+        ++(guard->count);
+    }
+
+    void unlock(wake_tbb_master wake) {
+        if (auto* p = guard.release()) {
+            wake_master(*p, wake);
+        }
+    }
+};
+
+master_thread_sleep_lock_t inline lock_sleep_master(async_tasks_t& async_tasks) {
+    return {&async_tasks};
+}
+
+enum class is_tbb_work_present {
+   no,
+   yes
```

Comment:
By coding style constants / enums should be upper case.
`NO` is a macro on some platforms (Apple).

Consider using `bool` type instead. Enum doesn't look as necessary here.

Reply:
using meaningful names  instead of faceless `true` / `false` , so `bool` is not an option :+1: 

will rename to UPPERCASE

Reply:
renamed